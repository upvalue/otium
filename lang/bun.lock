{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "lang",
      "dependencies": {
        "prettier": "^3.6.2",
        "raylib": "^0.14.0",
      },
      "devDependencies": {
        "@types/bun": "latest",
      },
      "peerDependencies": {
        "typescript": "^5",
      },
    },
  },
  "packages": {
    "@types/bun": ["@types/bun@1.2.20", "", { "dependencies": { "bun-types": "1.2.20" } }, "sha512-dX3RGzQ8+KgmMw7CsW4xT5ITBSCrSbfHc36SNT31EOUg/LA9JWq0VDdEXDRSe1InVWpd2yLUM1FUF/kEOyTzYA=="],

    "@types/node": ["@types/node@24.2.1", "", { "dependencies": { "undici-types": "~7.10.0" } }, "sha512-DRh5K+ka5eJic8CjH7td8QpYEV6Zo10gfRkjHCO3weqZHWDtAaSTFtl4+VMqOJ4N5jcuhZ9/l+yy8rVgw7BQeQ=="],

    "@types/react": ["@types/react@19.1.10", "", { "dependencies": { "csstype": "^3.0.2" } }, "sha512-EhBeSYX0Y6ye8pNebpKrwFJq7BoQ8J5SO6NlvNwwHjSj6adXJViPQrKlsyPw7hLBLvckEMO1yxeGdR82YBBlDg=="],

    "bun-types": ["bun-types@1.2.20", "", { "dependencies": { "@types/node": "*" }, "peerDependencies": { "@types/react": "^19" } }, "sha512-pxTnQYOrKvdOwyiyd/7sMt9yFOenN004Y6O4lCcCUoKVej48FS5cvTw9geRaEcB9TsDZaJKAxPTVvi8tFsVuXA=="],

    "cross-fetch": ["cross-fetch@3.2.0", "", { "dependencies": { "node-fetch": "^2.7.0" } }, "sha512-Q+xVJLoGOeIMXZmbUK4HYk+69cQH6LudR0Vu/pRm2YlU/hDV9CiS0gKUMaWY5f2NeUH9C1nV3bsTlCo0FsTV1Q=="],

    "csstype": ["csstype@3.1.3", "", {}, "sha512-M1uQkMl8rQK/szD0LNhtqxIPLpimGm8sOBwU7lLnCpSbTyY3yeU1Vc7l4KT5zT4s/yOxHH5O7tIuuLOCnLADRw=="],

    "module-alias": ["module-alias@2.2.3", "", {}, "sha512-23g5BFj4zdQL/b6tor7Ji+QY4pEfNH784BMslY9Qb0UnJWRAt+lQGLYmRaM0KDBwIG23ffEBELhZDP2rhi9f/Q=="],

    "node-fetch": ["node-fetch@2.7.0", "", { "dependencies": { "whatwg-url": "^5.0.0" }, "peerDependencies": { "encoding": "^0.1.0" }, "optionalPeers": ["encoding"] }, "sha512-c4FRfUm/dbcWZ7U+1Wq0AwCyFL+3nt2bEw05wfxSz+DWpWsitgmSgYmy2dQdWyKC1694ELPqMs/YzUSNozLt8A=="],

    "prettier": ["prettier@3.6.2", "", { "bin": { "prettier": "bin/prettier.cjs" } }, "sha512-I7AIg5boAr5R0FFtJ6rCfD+LFsWHp81dolrFD8S79U9tb8Az2nGrJncnMSnys+bpQJfRUzqs9hnA81OAA3hCuQ=="],

    "raylib": ["raylib@0.14.0", "", { "dependencies": { "cross-fetch": "^3.1.5", "module-alias": "^2.2.2" }, "bin": { "raylib": "bin/node-raylib" } }, "sha512-6dwsArtyWNhL+1R+uBpNlf+4Korpp24vdUKs24k27kMbvSmz8MJGh1ysXCjO5q6lQWKrfs/x4b2w+ygFV58I7g=="],

    "tr46": ["tr46@0.0.3", "", {}, "sha512-N3WMsuqV66lT30CrXNbEjx4GEwlow3v6rr4mCcv6prnfwhS01rkgyFdjPNBYd9br7LpXV1+Emh01fHnq2Gdgrw=="],

    "typescript": ["typescript@5.9.2", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-CWBzXQrc/qOkhidw1OzBTQuYRbfyxDXJMVJ1XNwUHGROVmuaeiEm3OslpZ1RV96d7SKKjZKrSJu3+t/xlw3R9A=="],

    "undici-types": ["undici-types@7.10.0", "", {}, "sha512-t5Fy/nfn+14LuOc2KNYg75vZqClpAiqscVvMygNnlsHBFpSXdJaYtXMcdNLpl/Qvc3P2cB3s6lOV51nqsFq4ag=="],

    "webidl-conversions": ["webidl-conversions@3.0.1", "", {}, "sha512-2JAn3z8AR6rjK8Sm8orRC0h/bcl/DqL7tRPdGZ4I1CjdF+EaMLmYxBHyXuKL849eucPFhvBoxMsflfOb8kxaeQ=="],

    "whatwg-url": ["whatwg-url@5.0.0", "", { "dependencies": { "tr46": "~0.0.3", "webidl-conversions": "^3.0.0" } }, "sha512-saE57nupxk6v3HY35+jzBwYa0rKSy0XR8JSxZPwgLr7ys0IBzhGviA1/TUGJLmSVqs8pb9AnvICXEuOHLprYTw=="],
  }
}
