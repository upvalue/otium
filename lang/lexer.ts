export class LexerError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "LexerError";
  }
}

export const TOKEN_TYPES = {
  NUMBER: "NUMBER",
  STRING: "STRING",
  SYMBOL: "SYMBOL",
  LPAREN: "LPAREN",
  RPAREN: "RPAREN",
  LBRACE: "LBRACE",
  RBRACE: "RBRACE",
  ADD: "ADD",
  SUB: "SUB",
  MUL: "MUL",
  DIV: "DIV",
  COMMA: "COMMA",
  ASSIGN: "ASSIGN",
  PUNCTUATION: "PUNCTUATION",
  EOF: "EOF",
  COMMENT: "COMMENT",
} as const;

export type Token = {
  type: (typeof TOKEN_TYPES)[keyof typeof TOKEN_TYPES];
  begin: number;
  end: number;
  value: string;
  sourceName?: string;
  line: number;
  column: number;
};

export const EOF_TOKEN: Token = {
  type: "EOF",
  begin: 0,
  end: 0,
  value: "",
  sourceName: "",
  line: 1,
  column: 1,
};

// characters which terminate symbols
const TERM_CHARS = ["(", ")", "[", "]", ".", '"', "'", "#", ","];

export class Lexer {
  private input: string;
  private position: number;
  private sourceName?: string;
  private lineNumber: number;
  private columnNumber: number;

  constructor(input: string, sourceName?: string) {
    this.input = input;
    this.position = 0;
    this.sourceName = sourceName;
    this.lineNumber = 1;
    this.columnNumber = 1;
  }

  private peek(): string {
    if (this.position >= this.input.length) {
      return "\0";
    }
    return this.input[this.position];
  }

  private advance(): string {
    const char = this.peek();
    if (char === "\n") {
      this.lineNumber++;
      this.columnNumber = 1;
    } else {
      this.columnNumber++;
    }
    this.position++;
    return char;
  }

  private skipWhitespace(): void {
    while (/\s/.test(this.peek())) {
      this.advance();
    }
  }

  private readNumber(): Token {
    const begin = this.position;
    const line = this.lineNumber;
    const column = this.columnNumber;

    while (/\d/.test(this.peek())) {
      this.advance();
    }

    return {
      type: "NUMBER",
      begin,
      end: this.position,
      value: this.input.slice(begin, this.position),
      sourceName: this.sourceName,
      line,
      column,
    };
  }

  private readString(): Token {
    const begin = this.position;
    const line = this.lineNumber;
    const column = this.columnNumber;
    const quote = this.advance(); // consume opening quote
    let value = "";

    while (this.peek() !== quote && this.peek() !== "\0") {
      if (this.peek() === "\\") {
        this.advance();
        const escaped = this.advance();
        switch (escaped) {
          case "n":
            value += "\n";
            break;
          case "t":
            value += "\t";
            break;
          case "r":
            value += "\r";
            break;
          case "\\":
            value += "\\";
            break;
          case '"':
            value += '"';
            break;
          case "'":
            value += "'";
            break;
          default:
            value += escaped;
        }
      } else {
        value += this.advance();
      }
    }

    if (this.peek() === quote) {
      this.advance(); // consume closing quote
    }

    return {
      type: "STRING",
      begin,
      end: this.position,
      value,
      sourceName: this.sourceName,
      line,
      column,
    };
  }

  private readSymbol(): Token {
    const begin = this.position;
    const line = this.lineNumber;
    const column = this.columnNumber;

    // First character must be alphabetic
    this.advance();

    let c = this.peek();

    // Rest can be alphanumeric or punctuation (but not whitespace or quotes)
    while (true) {
      c = this.peek();
      if (c !== "\0" && !/\s/.test(c) && !TERM_CHARS.includes(c)) {
        this.advance();
        continue;
      }
      break;
    }

    return {
      type: "SYMBOL",
      begin,
      end: this.position,
      value: this.input.slice(begin, this.position),
      sourceName: this.sourceName,
      line,
      column,
    };
  }

  nextToken(): Token {
    this.skipWhitespace();

    const { sourceName, lineNumber: line, columnNumber: column } = this;
    const begin = this.position;
    const char = this.peek();

    if (char === "\0") {
      return {
        type: "EOF",
        begin,
        end: this.position,
        value: "",
        sourceName,
        line,
        column,
      };
    }

    if (/\d/.test(char)) {
      return this.readNumber();
    }

    if (char === '"' || char === "'") {
      return this.readString();
    }

    if (/[a-zA-Z]/.test(char)) {
      return this.readSymbol();
    }

    // Check for specific punctuation
    this.advance();

    switch (char) {
      case "(":
        return {
          type: "LPAREN",
          begin,
          end: this.position,
          value: char,
          sourceName,
          line,
          column,
        };
      case ")":
        return {
          type: "RPAREN",
          begin,
          end: this.position,
          value: char,
          sourceName,
          line,
          column,
        };
      case ",": {
        return {
          type: "COMMA",
          begin,
          end: this.position,
          value: char,
          sourceName,
          line,
          column,
        };
      }
      case "{":
        return {
          type: "LBRACE",
          begin,
          end: this.position,
          value: char,
          sourceName,
          line,
          column,
        };
      case "}":
        return {
          type: "RBRACE",
          begin,
          end: this.position,
          value: char,
          sourceName,
          line,
          column,
        };
      case "+":
        return {
          type: "ADD",
          begin,
          end: this.position,
          value: char,
          sourceName,
          line,
          column,
        };
      case "-":
        return {
          type: "SUB",
          begin,
          end: this.position,
          value: char,
          sourceName,
          line,
          column,
        };
      case "*":
        return {
          type: "MUL",
          begin,
          end: this.position,
          value: char,
          sourceName,
          line,
          column,
        };
      case "/":
        return {
          type: "DIV",
          begin,
          end: this.position,
          value: char,
          sourceName,
          line,
          column,
        };
      case ":":
        if (this.peek() == "=") {
          this.advance();
          return {
            type: "ASSIGN",
            begin,
            end: this.position,
            value: ":=",
            sourceName,
            line: line,
            column: column,
          };
        }
      default:
        return {
          type: "PUNCTUATION",
          begin,
          end: this.position,
          value: char,
          sourceName,
          line,
          column,
        };
    }
  }

  tokenize(): Token[] {
    const tokens: Token[] = [];
    let token = this.nextToken();

    while (token.type !== "EOF") {
      tokens.push(token);
      token = this.nextToken();
    }

    tokens.push(token); // Include EOF token
    return tokens;
  }
}

if (import.meta.main) {
  const args = process.argv.slice(2);
  if (args.length !== 1) {
    console.error("Usage: bun lexer.ts <file>");
    process.exit(1);
  }

  const filename = args[0];

  try {
    const fs = require("fs");
    const content = fs.readFileSync(filename, "utf8");
    const lexer = new Lexer(content);

    let token = lexer.nextToken();
    while (token.type !== "EOF") {
      console.log(`${token.type} ${token.line}:${token.column} ${token.value}`);
      token = lexer.nextToken();
    }
    console.log(`${token.type} ${token.value}`);
  } catch (error) {
    console.error(`Error reading file: ${error}`);
    process.exit(1);
  }
}
