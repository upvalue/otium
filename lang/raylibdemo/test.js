///// otium:prelude
const print = console.log;

// Builtin operators
const _ot_lt = (a, b) => a < b;
const _ot_sub = (a, b) => a - b;
const _ot_add = (a, b) => a + b;
const _ot_lte = (a, b) => a <= b;
const _ot_gte = (a, b) => a >= b;
const _ot_eq = (a, b) => a === b;
const _ot_neq = (a, b) => a !== b;
const _ot_mul = (a, b) => a * b;
const _ot_div = (a, b) => a / b;

const not = (a) => !a;

///// otium:main

// scratch.ot 4:1
/* r := js("require('raylib')") */
let r, def0;
def0 = require("raylib");
r = def0;
// scratch.ot 6:1
/* screenWidth := 800 */
let screenWidth, def1;
def1 = 800;
screenWidth = def1;
// scratch.ot 7:1
/* screenHeight := 600 */
let screenHeight, def2;
def2 = 600;
screenHeight = def2;
// scratch.ot 9:1
/* ballPos := r.Vector2(400, 300) */
let ballPos, def3;
let apply_operator4, apply_operand5_0, apply_operand6_1;
apply_operator4 = r.Vector2;
apply_operand5_0 = 400;
apply_operand6_1 = 300;

def3 = apply_operator4(apply_operand5_0, apply_operand6_1);
ballPos = def3;
// scratch.ot 11:1
/* r.InitWindow(screenWidth, screenHeight, "raylib ball movement demo in otium") */
let apply_operator7, apply_operand8_0, apply_operand9_1, apply_operand10_2;
apply_operator7 = r.InitWindow;
apply_operand8_0 = screenWidth;
apply_operand9_1 = screenHeight;
apply_operand10_2 = "raylib ball movement demo in otium";

apply_operator7(apply_operand8_0, apply_operand9_1, apply_operand10_2);
// scratch.ot 12:1
/* r.SetTargetFPS(60) */
let apply_operator11, apply_operand12_0;
apply_operator11 = r.SetTargetFPS;
apply_operand12_0 = 60;

apply_operator11(apply_operand12_0);
// scratch.ot 14:1
/* while(not(r.WindowShouldClose())) {
  r.BeginDrawing()

  if(r.IsKeyDown(r.KEY_RIGHT), ballPos.x = ballPos.x + 2)
  if(r.IsKeyDown(r.KEY_LEFT), ballPos.x = ballPos.x - 2)
  if(r.IsKeyDown(r.KEY_UP), ballPos.y = ballPos.y - 2)
  if(r.IsKeyDown(r.KEY_DOWN), ballPos.y = ballPos.y + 2)

  r.ClearBackground(r.RAYWHITE)
  r.DrawText("move the ball with arrow keys", 10, 10, 20, r.DARKGRAY)
  r.DrawCircleV(ballPos, 50, r.MAROON)

  r.EndDrawing()
} */
let while_cond13;
let apply_operator15, apply_operand16_0;
apply_operator15 = not;
let apply_operator17;
apply_operator17 = r.WindowShouldClose;

apply_operand16_0 = apply_operator17();

while_cond13 = apply_operator15(apply_operand16_0);
while (while_cond13 !== false) {
  let apply_operator18;
  apply_operator18 = r.BeginDrawing;

  apply_operator18();
  let if_cond19;
  let apply_operator20, apply_operand21_0;
  apply_operator20 = r.IsKeyDown;
  apply_operand21_0 = r.KEY_RIGHT;

  if_cond19 = apply_operator20(apply_operand21_0);
  if (if_cond19 !== false) {
    let apply_operator23, apply_operand24_0, apply_operand25_1;
    apply_operator23 = _ot_add;
    apply_operand24_0 = ballPos.x;
    apply_operand25_1 = 2;

    set22 = apply_operator23(apply_operand24_0, apply_operand25_1);
    ballPos.x = set22;
  }
  let if_cond26;
  let apply_operator27, apply_operand28_0;
  apply_operator27 = r.IsKeyDown;
  apply_operand28_0 = r.KEY_LEFT;

  if_cond26 = apply_operator27(apply_operand28_0);
  if (if_cond26 !== false) {
    let apply_operator30, apply_operand31_0, apply_operand32_1;
    apply_operator30 = _ot_sub;
    apply_operand31_0 = ballPos.x;
    apply_operand32_1 = 2;

    set29 = apply_operator30(apply_operand31_0, apply_operand32_1);
    ballPos.x = set29;
  }
  let if_cond33;
  let apply_operator34, apply_operand35_0;
  apply_operator34 = r.IsKeyDown;
  apply_operand35_0 = r.KEY_UP;

  if_cond33 = apply_operator34(apply_operand35_0);
  if (if_cond33 !== false) {
    let apply_operator37, apply_operand38_0, apply_operand39_1;
    apply_operator37 = _ot_sub;
    apply_operand38_0 = ballPos.y;
    apply_operand39_1 = 2;

    set36 = apply_operator37(apply_operand38_0, apply_operand39_1);
    ballPos.y = set36;
  }
  let if_cond40;
  let apply_operator41, apply_operand42_0;
  apply_operator41 = r.IsKeyDown;
  apply_operand42_0 = r.KEY_DOWN;

  if_cond40 = apply_operator41(apply_operand42_0);
  if (if_cond40 !== false) {
    let apply_operator44, apply_operand45_0, apply_operand46_1;
    apply_operator44 = _ot_add;
    apply_operand45_0 = ballPos.y;
    apply_operand46_1 = 2;

    set43 = apply_operator44(apply_operand45_0, apply_operand46_1);
    ballPos.y = set43;
  }
  let apply_operator47, apply_operand48_0;
  apply_operator47 = r.ClearBackground;
  apply_operand48_0 = r.RAYWHITE;

  apply_operator47(apply_operand48_0);
  let apply_operator49,
    apply_operand50_0,
    apply_operand51_1,
    apply_operand52_2,
    apply_operand53_3,
    apply_operand54_4;
  apply_operator49 = r.DrawText;
  apply_operand50_0 = "move the ball with arrow keys";
  apply_operand51_1 = 10;
  apply_operand52_2 = 10;
  apply_operand53_3 = 20;
  apply_operand54_4 = r.DARKGRAY;

  apply_operator49(
    apply_operand50_0,
    apply_operand51_1,
    apply_operand52_2,
    apply_operand53_3,
    apply_operand54_4,
  );
  let apply_operator55, apply_operand56_0, apply_operand57_1, apply_operand58_2;
  apply_operator55 = r.DrawCircleV;
  apply_operand56_0 = ballPos;
  apply_operand57_1 = 50;
  apply_operand58_2 = r.MAROON;

  apply_operator55(apply_operand56_0, apply_operand57_1, apply_operand58_2);
  let apply_operator59;
  apply_operator59 = r.EndDrawing;

  while_body14 = apply_operator59();
  let apply_operator60, apply_operand61_0;
  apply_operator60 = not;
  let apply_operator62;
  apply_operator62 = r.WindowShouldClose;

  apply_operand61_0 = apply_operator62();

  while_cond13 = apply_operator60(apply_operand61_0);
}
// scratch.ot 29:1
/* r.CloseWindow() */
let apply_operator63;
apply_operator63 = r.CloseWindow;

apply_operator63();
// scratch.ot 31:1
/*  */

