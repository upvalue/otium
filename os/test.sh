#!/usr/bin/env bash
set -euxo pipefail
clang++ -DOT_TEST -DOT_TRACE_MEM \
    -I. -o test \
    otk/memory-test.cpp \
    otk/memory.cpp \
    otk/std.cpp \
    otk/platform-test.cpp 
./test