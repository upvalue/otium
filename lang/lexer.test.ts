import { expect, test } from "bun:test";
import { Lexer } from "./lexer";

test("basic lexer test", () => {
    const input = `42 "hello world" foo-bar ( ) + - * / ;`;
    const lexer = new Lexer(input);
    const tokens = lexer.tokenize();
    
    expect(tokens).toMatchInlineSnapshot(`
      [
        {
          "begin": 0,
          "end": 2,
          "type": "NUMBER",
          "value": "42",
        },
        {
          "begin": 3,
          "end": 16,
          "type": "STRING",
          "value": "hello world",
        },
        {
          "begin": 17,
          "end": 24,
          "type": "SYMBOL",
          "value": "foo-bar",
        },
        {
          "begin": 25,
          "end": 26,
          "type": "LPAREN",
          "value": "(",
        },
        {
          "begin": 27,
          "end": 28,
          "type": "RPAREN",
          "value": ")",
        },
        {
          "begin": 29,
          "end": 30,
          "type": "ADD",
          "value": "+",
        },
        {
          "begin": 31,
          "end": 32,
          "type": "SUB",
          "value": "-",
        },
        {
          "begin": 33,
          "end": 34,
          "type": "MUL",
          "value": "*",
        },
        {
          "begin": 35,
          "end": 36,
          "type": "DIV",
          "value": "/",
        },
        {
          "begin": 37,
          "end": 38,
          "type": "PUNCTUATION",
          "value": ";",
        },
        {
          "begin": 38,
          "end": 38,
          "type": "EOF",
          "value": "",
        },
      ]
    `);
});