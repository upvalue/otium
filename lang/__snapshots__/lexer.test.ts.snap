// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`basic lexer test 1`] = `
[
  {
    "begin": 0,
    "end": 2,
    "type": "NUMBER",
    "value": "42",
  },
  {
    "begin": 3,
    "end": 16,
    "type": "STRING",
    "value": "hello world",
  },
  {
    "begin": 17,
    "end": 24,
    "type": "SYMBOL",
    "value": "foo-bar",
  },
  {
    "begin": 25,
    "end": 26,
    "type": "PUNCTUATION",
    "value": "(",
  },
  {
    "begin": 27,
    "end": 28,
    "type": "PUNCTUATION",
    "value": ")",
  },
  {
    "begin": 29,
    "end": 30,
    "type": "PUNCTUATION",
    "value": ";",
  },
  {
    "begin": 30,
    "end": 30,
    "type": "EOF",
    "value": "",
  },
]
`;
